src/main/java/hudson/plugins/gradle/injection/MavenExtensionsHandler.java

```diff
@@ -20,39 +20,36 @@ public class MavenExtensionsHandler {
     private final MavenExtensionFileHandler geExtensionHandler = new MavenExtensionFileHandler(GRADLE_ENTERPRISE);
     private final MavenExtensionFileHandler ccudExtensionHandler = new MavenExtensionFileHandler(CCUD);
 
-    public void copyGradleEnterpriseExtensionToAgent(FilePath rootPath) throws IOException, InterruptedException {
-        geExtensionHandler.copyExtensionToAgent(rootPath);
+    public FilePath copyGradleEnterpriseExtensionToAgent(FilePath rootPath) throws IOException, InterruptedException {
+        return geExtensionHandler.copyExtensionToAgent(rootPath);
     }
 
-    public FilePath getGradleEnterpriseExtensionPath(FilePath rootPath) throws IOException {
-        return geExtensionHandler.getAgentExtensionPath(rootPath);
-    }
-
-    public void copyCCUDExtensionToAgent(FilePath rootPath) throws IOException, InterruptedException {
-        ccudExtensionHandler.copyExtensionToAgent(rootPath);
-    }
-
-    public FilePath getCCUDExtensionPath(FilePath rootPath) throws IOException {
-        return ccudExtensionHandler.getAgentExtensionPath(rootPath);
+    public FilePath copyCCUDExtensionToAgent(FilePath rootPath) throws IOException, InterruptedException {
+        return ccudExtensionHandler.copyExtensionToAgent(rootPath);
     }
 
     public void deleteAllExtensionsFromAgent(FilePath rootPath) throws IOException, InterruptedException {
         rootPath.child(LIB_DIR_PATH).deleteContents();
     }
 
     private static final class MavenExtensionFileHandler {
+
         private final MavenExtension extension;
 
         MavenExtensionFileHandler(MavenExtension extension) {
             this.extension = extension;
         }
 
-        public void copyExtensionToAgent(FilePath rootPath) throws IOException, InterruptedException {
-            copyResourceToNode(rootPath.child(LIB_DIR_PATH).child(extension.getJarName()), extension.getJarName());
-        }
-
-        public FilePath getAgentExtensionPath(FilePath rootPath) {
-            return rootPath.child(LIB_DIR_PATH).child(extension.getJarName());
+        /**
+         * Copies the extension to the agent, if it is not already present, and returns a path to the extension
+         * on the agent.
+         */
+        public FilePath copyExtensionToAgent(FilePath rootPath) throws IOException, InterruptedException {
+            FilePath nodePath = rootPath.child(LIB_DIR_PATH).child(extension.getJarName());
+            if (!nodePath.exists()) {
```

Comment:
Old versions will not be cleaned up. I think we should always clean them up. Please use the same logic as for Gradle. You can just use `gradle-enterprise-maven-extension.jar` for the jar name, this way you don't have to delete other files and you can just replace it if it changed or keep the version in the name and delete it, however you prefer.

Reply:
To keep things simple, I decided to rely on cleanup during the disabling. But I agree, it's a tradeoff. I think that dealing with versions is harder. After the upgrade, we know the new version, but don't know what the previous version was. It will be more code to clean it. Using files without versions with checksum is simpler, because, yes, we can just replace files.

Reply:
Changed.